src/spec/sync/product-sync.spec.coffee

```diff
@@ -109,13 +109,186 @@ NEW_PRODUCT =
+      expected_update =
+        actions: [
+          {
+            sku: 'sku3',
+            action: 'addVariant'
+          }, {
+            action: 'changeMasterVariant',
+            sku: 'sku3'
+          }, {
+            action: 'removeVariant',
+            id: 1
+          }
+        ]
+        version: oldProduct.version
+      expect(update).toEqual expected_update
+
+    it 'should change the masterVariant', ->
+      oldProduct =
+        id: '123'
+        version: 1
+        masterVariant:
+          id: 1
+          sku: 'sku1'
+        variants: [
+          { id: 2, sku: 'sku2' }
+        ]
+
+      newProduct =
+        id: '123'
+        version: 1
+        masterVariant:
+          id: 2
+          sku: 'sku2'
+        variants: [
+          { id: 1, sku: 'sku1' }
+        ]
+
+      update = @sync.buildActions(newProduct, oldProduct).getUpdatePayload()
+      expected_update =
+        actions: [
+          { action: 'changeMasterVariant', variantId: 2 }
+        ]
+        version: oldProduct.version
+      expect(update).toEqual expected_update
+
+    it 'should not do any action when variant order changed when id is provided', ->
+      oldProduct =
+        id: '123'
+        version: 1
+        masterVariant:
+          id: 1
+        variants: [
+          { id: 2 }
+          { id: 3 }
+        ]
+
+      newProduct =
+        id: '123'
+        version: 1
+        masterVariant:
+          id: 1
+        variants: [
+          { id: 3 }
+          { id: 2 }
+        ]
+
+      update = @sync.buildActions(newProduct, oldProduct).getUpdatePayload()
+      expect(update).toEqual undefined
+
+    it 'should not do any action when variant order changed when sku is provided', ->
+      oldProduct =
+        id: '123'
+        version: 1
+        masterVariant:
+          id: 1
+          sku: 'sku1'
+        variants: [
+          { id: 2, sku: 'sku2' }
+          { id: 3, sku: 'sku3' }
+        ]
+
+      newProduct =
+        id: '123'
+        version: 1
+        masterVariant:
+          sku: 'sku1'
+        variants: [
+          { sku: 'sku3' }
+          { sku: 'sku2' }
+        ]
+
+      update = @sync.buildActions(newProduct, oldProduct).getUpdatePayload()
+      # no action
+      expect(update).toEqual undefined
+
+    it 'should throw an error when no id or sku is provided', ->
+      oldProduct =
+        id: '123'
+        version: 1
+        masterVariant:
+          id: 1
+          key: 'sku1'
+        variants: [
+          { id: 2, key: 'sku2' }
+          { id: 3, key: 'sku3' }
+        ]
+
+      newProduct =
+        id: '123'
+        version: 1
+        masterVariant:
+          key: 'sku1'
+        variants: [
+          { key: 'sku3' }
+          { key: 'sku2' }
+        ]
+
+      expect(=> @sync.buildActions(newProduct, oldProduct)).toThrow new Error 'A variant must either have an ID or an SKU.'
+
+#    it 'should handle removing variant', ->
```

Comment:
What about these tests?

Reply:
I removed them here 01fd315.. the functionality of adding and removing variants is already tested in other tests